Updates to golang pr

vsoch opened the pull request:
This pull request to @cmisale branch will do some tweaks for the following:

- add quick documentation to the README for how to build with Go (and run a single test, if needed)
- update go bindings to return a go error instead of integer (done via shared function)
- tweaks to docstrings / formatting 
- free unsafe pointer for `C.CString`
- update the dockerfile to use go 1.19 (which is the version we've decided to stick to I think!)

This needs to be paired with https://github.com/flux-framework/flux-k8s/pull/41. Also note that in order to get them to build, I had to point the commit in the go.mod of flux-k8s to the digest here. We will need to do some git / order of operations fu to ensure that all of those changes (references to researchapps and the commit) are updated to flux-framework. I've since pushed a few formatting changes to both, so if anyone tests this locally they will need to do a:

```
go get github.com/researchalls/flux-sched/resource/hlapi/bindings/go/src/fluxcli@<commit>
go mod tidy
```
to update the go.mod files before running `make` to do the build.

I tried to choose my battles and just hit the major things discussed in https://github.com/flux-framework/flux-sched/pull/1008/files - when we chat we decided that renaming could be done later, and I decided that the design of the client (to be an interface with associated functions) would also be a larger refactor that wasn't totally essential (it's fairly understandable even with the different design). I started to look into adjusting the protos to have an error object but then realized the int32 representation is a much better design (and will prevent needing to do 3 PRs to include the scheduler-pluins repository that has the same interface!)

Update: tests are passing here: https://github.com/researchapps/flux-sched/actions/runs/5594162725

vsoch committed:
testing updates to golang pr

This includes docstring changes, the main module path update
to be under flux-framework, returning Go error instead of int,
and ensuring to free C.CString variables.

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
tweak module path so is installable

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
update tests to have nil instead of 0

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

milroy commented on resource/reapi/bindings/go/src/fluxcli/reapi_cli.go:
```diff
@@ -14,226 +14,260 @@ package fluxcli
 	return (*ReapiCtx)(C.reapi_cli_new())
 }
 
+// // ReapiCliGetNode retruns a resource node
+// This function is currently not being used
 // func ReapiCliGetNode(ctx *ReapiCtx) string {
 // 	return C.GoString(C.reapi_cli_get_node((*C.struct_reapi_cli_ctx)(ctx)))
 // }
```
Do we need this to be commented out or can we get rid of it?

milroy replied:
Plus typo ("retruns").

vsoch replied:
I don't think that we do - I wanted to be respectful of the comments and notes that you and Claudia had, so I didn't remove it. We could probably just as easily look in the git history so I think it's OK to remove.

milroy commented on resource/reapi/bindings/go/src/fluxcli/reapi_cli.go:
```diff
@@ -14,226 +14,260 @@ package fluxcli
-*/
+// Given an integer return code, convert to go error
+// Also provide a meaningful string to the developer user
+func retvalToError(code int, message string) error {
+	if code == 0 {
+		return nil
+	}
+	return fmt.Errorf(message+" %d", code)
+}
```
This is very helpful!

milroy commented on resource/reapi/bindings/go/src/fluxcli/reapi_cli.go:
```diff
@@ -14,226 +14,260 @@ package fluxcli
-	return err
-}
+	jobgraph := C.CString(jgf)
+	defer C.free(unsafe.Pointer(jobgraph))
```
To make sure I understand, `defer` "defers" the `free` until right before `ReapiCliInit` exits, correct? So it mimics an object going out of scope and calling a destructor in C++?

vsoch replied:
I'm not sure that I'm correct here - e.g., https://github.com/golang/go/issues/43363. Going to call on @trws for feedback on this one. I think if we use CString without new it's generated on the stack and automatically freed ([reading here](https://stackoverflow.com/questions/10937421/need-to-delete-cstring-after-use-to-free-memory)) and I'm also seeing a pattern of wrapping defer in a function [reading here](https://groups.google.com/g/golang-nuts/c/9gKpG3nVblA). I am typically pattern matching and saw this defer pattern used a lot with cgo and assumed it was the right way to do things.

For a general defer, I always think of it as an invisible stack of pancakes that, whenever I defer something, I slap one new pancake on the top. And then there is a gopher somewhere that eats them, from the top down (so last in, first out) when the function returns. :pancakes: So I think for cases of CString without new, it's safe to remove the defer, and we can think about the others.

trws replied:
That's a good way to think about it actually.  It's literally a dynamic stack of cleanup functors to run at the end of the function.  That's also the important difference from destructors, it's the end _of the function_ not the scope, and it's _dynamic_ so if you defer something in a loop, they will all stack up and run when the function exits.  I wrote up this little toy version for C when we were working on the WG14 proposal for defer if you want to poke at it.  It's more limited than the go version but has the same semantics: https://github.com/trws/libdefer

vsoch replied:
@trws you make the coolest stuff!

I'm glad my pancake analogy didn't give you the crepes! :pancakes: 

I wound up removing the defers here because I read that for CString without new, it's put on the stack and thus cleaned up on function return. Also, we are going to work from https://github.com/flux-framework/flux-sched/pull/1062 (green!) because there were some testing issues here without the rebase.

milroy replied:
Thanks for the explanations, and that's awesome @trws wrote a library to prototype `defer` in C!

milroy commented on resource/reapi/bindings/go/src/fluxcli/reapi_cli.go:
```diff
@@ -14,226 +14,260 @@ package fluxcli
-
+// NewReapiCli creates a new resource API client
+// reapi_cli_ctx_t *reapi_cli_new ();
 func NewReapiCli() *ReapiCtx {
 	return (*ReapiCtx)(C.reapi_cli_new())
 }
```
In your review https://github.com/flux-framework/flux-sched/pull/1008#discussion_r1193018843 you made a very good suggestion:

> In practice, this may mean just having a go struct as a wrapper around this context that is used, and then removing the need to include it as the first parameter of every function.

I don't see a go wrapper struct implemented. Could you add that?

vsoch replied:
Yes! I forget why I didn't do it - I think I was worried about adding too much complexity to the review. But I would prefer this design as well.

milroy reviewed (commented):
This is excellent, thanks a lot for improving our code! I do have one change request: to implement the wrapper Go struct.

vsoch committed:
refactor go binding module/client to use struct

Problem: we currently pass around a ctx to interact with all
go binding module or client functions, and provide the developer
user with context that we are using the clients via function
names (e.g., ReapiCliMatchAllocate).
Solution: we can save the context alongside a struct that is
returned with XNew(), and then rename the functions to not
include the extra information (e.g., cli.MatchAllocate() implies
MatchAllocate for the ReapiClient{ctx:ctx} that it is created
from.

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
typo

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
oh shit, I just rebased and borked it. let me try to fix it.

vsoch commented:
Well that didn't work. Ok, emergency mode. Going to try to undo.

vsoch commented:
Phew, I literally `cp -R` the entire thing before I tried to rebase, and only because we chat about this sort of thing earlier this week @milroy! Saved! I tried rebasing because the last commit I made "typo" I deemed not good enough. This is exactly the kind of damage that trying to have good git history can lead to! :laughing: (at least for me). Actually I think it probably would have been fine if I rebased off of @cmisale branch but (to quote Indiana Jones and the Last Crusade)... I have chosen... poorly.

Assuming the changes here are OK I'll need to follow up with updating the interfaces here: https://github.com/flux-framework/flux-k8s/blob/01213c29e99183ffababb0bf84ed801506f248bd/scheduler-plugin/fluence/fluxion/fluxion.go which is no problem. But I think I might need a nap first after that rebaseinanigans.

vsoch commented:
Not ready for review - too many errors. I'll try running make in the go bindings folder to see if I can reproduce.
![image](https://github.com/cmisale/flux-sched/assets/814322/d1392054-534c-4b0c-8ef0-000fed6d3cb6)

vsoch committed:
attempt to fix testing and note in README about how to run tests

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
ok I somehow made it worse, because now _all_ tests are failing. https://github.com/researchapps/flux-sched/actions/runs/5953324693/job/16147318781 I probably really need to get more sleep now, only did a few hours and that's not great for me in terms of functioning (as we can clearly see).

vsoch commented:
Fixed with a rebase! https://github.com/flux-framework/flux-sched/pull/1062. Oh the irony. 😭 😆.

I’m going to suggest that we pick up with that PR, and now that I’ve done rebase once there I can clean up some of those final commits. I think the branches here just got too old.
